File: other/cracking-the-coding-interview/ch1/1.2_check_permutation/1_2_check_permutation.go
package main

import (
	"fmt"
	"sort"
)

func isPermutation(input1, input2 string) bool {
	if len(input1) != len(input2) {
		return false
	}
	counts := make(map[rune]int)
	for _, r := range input1 {
		counts[r]++
	}
	for _, r := range input2 {
		counts[r]--
	}
	for _, val := range counts {
		if val != 0 {
			return false
		}
	}
	return true
}

// by sort
func isPermutation2(input1, input2 string) bool {
	s1 := []rune(input1)
	sort.Slice(s1, func(i int, j int) bool { return s1[i] < s1[j] })

	s2 := []rune(input2)
	sort.Slice(s2, func(i int, j int) bool { return s2[i] < s2[j] })

	return !(string(s1) != string(s2))
}

func main() {
	inputs1 := []string{"testest", "hello", "", ""}
	inputs2 := []string{"estxest", "oellh", "", "abc"}
	outputs := []bool{false, true, true, false}
	var sol bool
	var ans bool

	for i := 0; i < len(outputs); i++ {
		sol = isPermutation(inputs1[i], inputs2[i])
		ans = outputs[i]

		if sol != ans {
			fmt.Println(inputs1[i], inputs2[i], sol, ans)
		}
	}
}

File: other/cracking-the-coding-interview/ch1/1.2_check_permutation/1_2_check_permutation.py

class Solution:
    def is_permutation(self, s1, s2):
        if len(s1) != len(s2):
            return False

        count = {}
        for s in s1:
            if s in count:
                count[s] += 1
            else:
                count[s] = 1

        for s in s2:
            if s in count:
                count[s] -= 1
            else:
                count[s] = -1

        for v in count.values():
            if v != 0:
                return False

        return True


if __name__ == '__main__':
    inputs1 = ["testest", "hello", "", ""]
    inputs2 = ["estxest", "oellh", "", "abc"]
    outputs = [False, True, True, False]
    sol = Solution()

    for i in range(len(outputs)):
        if sol.is_permutation(inputs1[i], inputs2[i]) != outputs[i]:
            print(inputs1[i],  inputs2[i], outputs[i],
                  sol.is_permutation(inputs1[i], inputs2[i],))
